src/App.tsx

```diff
@@ -15,36 +15,99 @@ export const goodsFromServer = [
+enum SortType {
+  Default = 'default',
+  Alphabetical = 'alphabetical',
+  Length = 'length',
+}
+
 export const App: React.FC = () => {
+  const [goods, setGoods] = useState<string[]>([...goodsFromServer]);
+  const [isReversed, setIsReversed] = useState<boolean>(false);
+  const [sortType, setSortType] = useState<SortType>(SortType.Default);
+
+  const handleSortAlphabetically = () => {
+    const sortedGoods = [...goodsFromServer].sort((a, b) => a.localeCompare(b));
+
+    setGoods(sortedGoods);
+    setSortType(SortType.Alphabetical);
+    setIsReversed(false);
+  };
+
+  const handleSortByLength = () => {
+    const sortedGoods = [...goodsFromServer].sort(
+      (a, b) => a.length - b.length,
+    );
+
+    setGoods(sortedGoods);
+    setSortType(SortType.Length);
+    setIsReversed(false);
+  };
+
+  const handleReverseGoods = () => {
+    setGoods(prevGoods => [...prevGoods].reverse());
+    setIsReversed(prev => !prev);
+  };
+
+  const handleResetGoods = () => {
+    setGoods([...goodsFromServer]);
+    setSortType(SortType.Default);
+    setIsReversed(false);
+  };
+
+  const isOriginalOrder =
+    JSON.stringify(goods) === JSON.stringify(goodsFromServer);
+
   return (
     <div className="section content">
       <div className="buttons">
-        <button type="button" className="button is-info is-light">
+        <button
+          type="button"
+          className={`button is-info ${
+            sortType === SortType.Alphabetical && !isReversed ? '' : 'is-light'
+          }`}
+          onClick={handleSortAlphabetically}
+        >
           Sort alphabetically
         </button>
 
-        <button type="button" className="button is-success is-light">
+        <button
+          type="button"
+          className={`button is-success ${
+            sortType === SortType.Length && !isReversed ? '' : 'is-light'
+          }`}
+          onClick={handleSortByLength}
+        >
           Sort by length
         </button>
 
-        <button type="button" className="button is-warning is-light">
+        <button
+          type="button"
+          className={`button is-warning ${isReversed ? '' : 'is-light'}`}
+          onClick={handleReverseGoods}
+        >
           Reverse
         </button>
 
-        <button type="button" className="button is-danger is-light">
-          Reset
-        </button>
+        {!isOriginalOrder && (
+          <button
+            type="button"
+            className="button is-danger"
+            onClick={handleResetGoods}
+          >
+            Reset
+          </button>
+        )}
       </div>
 
       <ul>
-        <ul>
-          <li data-cy="Good">Dumplings</li>
-          <li data-cy="Good">Carrot</li>
-          <li data-cy="Good">Eggs</li>
-          <li data-cy="Good">Ice cream</li>
-          <li data-cy="Good">Apple</li>
-          <li data-cy="Good">...</li>
-        </ul>
+        {goods.map((good, index) => (
+          <li key={index} data-cy="Good">
```

Comment:
Using the index as a key in list rendering (`key={index}`) can lead to issues if the list items change dynamically. It's better to use a unique identifier for each item if possible. If the items are static and won't change, this approach is acceptable.